dpnp/dpnp_algo/dpnp_elementwise_common.py

```diff
@@ -24,89 +24,453 @@
+            sycl_dev - The :class:`dpctl.SyclDevice` where the function
+                evaluation is carried out.
+        The function is invoked when the argument of the unary function
+        requires casting, e.g. the argument of `dpctl.tensor.log` is an
+        array with integral data type.
+    """
+
+    def __init__(
+        self,
+        name,
+        result_type_resolver_fn,
+        unary_dp_impl_fn,
+        docs,
+        origin_fn=None,
+        mkl_fn_to_call=None,
+        mkl_impl_fn=None,
+        acceptance_fn=None,
+    ):
+        def _call_func(src, dst, sycl_queue, depends=None):
+            """A callback to register in UnaryElementwiseFunc class of dpctl.tensor"""
+
+            if depends is None:
+                depends = []
+
+            if mkl_fn_to_call is not None and mkl_fn_to_call(
+                sycl_queue, src, dst
+            ):
+                # call pybind11 extension for unary function from OneMKL VM
+                return mkl_impl_fn(sycl_queue, src, dst, depends)
+            return unary_dp_impl_fn(src, dst, sycl_queue, depends)
+
+        super().__init__(
+            name,
+            result_type_resolver_fn,
+            _call_func,
+            docs,
+            acceptance_fn=acceptance_fn,
+        )
+        self.__name__ = "DPNPUnaryFunc"
+        self.origin_fn = origin_fn
+
+    def __call__(
+        self,
+        x,
+        out=None,
+        where=True,
+        order="K",
+        dtype=None,
+        subok=True,
+        **kwargs,
+    ):
+        if kwargs:
+            pass
+        elif where is not True:
+            pass
+        elif dtype is not None:
+            pass
+        elif subok is not True:
+            pass
+        elif dpnp.isscalar(x):
+            # input has to be an array
+            pass
+        else:
+            if order in "afkcAFKC":
+                order = order.upper()
+            elif order is None:
+                order = "K"
+            else:
+                raise ValueError(
+                    "order must be one of 'C', 'F', 'A', or 'K' (got '{}')".format(
+                        order
+                    )
+                )
+            x_usm = dpnp.get_usm_ndarray(x)
+            out_usm = None if out is None else dpnp.get_usm_ndarray(out)
+            res_usm = super().__call__(x_usm, out=out_usm, order=order)
+            if out is not None and isinstance(out, dpnp_array):
+                return out
+            return dpnp_array._create_from_usm_ndarray(res_usm)
+
+        if self.origin_fn is not None:
+            return call_origin(
+                self.origin_fn,
+                x,
+                out=out,
+                where=where,
+                order=order,
+                dtype=dtype,
+                subok=subok,
+                **kwargs,
+            )
+        else:
+            raise NotImplementedError(
+                f"Requested function={self.name_} with args={x} and kwargs={kwargs} "
```

Comment:
Wouldn't it better to clear state what exact argument is out of supported range?